packages/backend-cli/src/directLine.ts

```diff
@@ -0,0 +1,772 @@
+          if (error.status === 403) {
+            // if the token is expired there's no reason to keep trying
+            this.expiredToken();
+            return Observable.throw(error);
+          }
+          return Observable.of(error);
+        })
+        .delay(timeout)
+        .take(retries)
+      )
+    )
+  }
+
+  public reconnect(conversation: Conversation) {
+    this.token = conversation.token;
+    this.streamUrl = conversation.streamUrl;
+    if (this.connectionStatus$.getValue() === ConnectionStatus.ExpiredToken)
+      this.connectionStatus$.next(ConnectionStatus.Online);
+  }
+
+  end() {
+    if (this.tokenRefreshSubscription)
+      this.tokenRefreshSubscription.unsubscribe();
+    this.connectionStatus$.next(ConnectionStatus.Ended);
+  }
+
+  getSessionId(): Observable<string> {
+    // If we're not connected to the bot, get connected
+    // Will throw an error if we are not connected
+    konsole.log("getSessionId");
+    return this.checkConnection(true)
+      .flatMap(_ =>
+        this.fetchAsObservable({
+          method: "GET",
+          url: `${this.domain}/session/getsessionid`,
+          withCredentials: true,
+          timeout,
+          headers: {
+            "Content-Type": "application/json",
+            "Authorization": `Bearer ${this.token}`
+          }
+        })
+        .map(ajaxResponse => {
+          konsole.log("getSessionId response: " + ajaxResponse.response.sessionId);
+          return ajaxResponse.response.sessionId as string;
+        })
+        .catch(error => this.catchPostError(error))
+      )
+      .catch(error => this.catchExpiredToken(error));
+  }
+
+  postActivity(activity: Activity) {
+    // Use postMessageWithAttachments for messages with attachments that are local files (e.g. an image to upload)
+    // Technically we could use it for *all* activities, but postActivity is much lighter weight
+    // So, since WebChat is partially a reference implementation of Direct Line, we implement both.
+    if (activity.type === "message" && activity.attachments && activity.attachments.length > 0)
+      return this.postMessageWithAttachments(activity);
+
+    // If we're not connected to the bot, get connected
+    // Will throw an error if we are not connected
+    konsole.log("postActivity", activity);
+    return this.checkConnection(true)
+    .flatMap(_ =>
+      this.fetchAsObservable({
+        method: "POST",
+        url: `${this.domain}/conversations/${this.conversationId}/activities`,
+        body: JSON.stringify(activity),
+        timeout,
+        headers: {
+          "Content-Type": "application/json",
+          "Authorization": `Bearer ${this.token}`
+        }
+      })
+      .map(ajaxResponse => ajaxResponse.response.id as string)
+      .catch(error => this.catchPostError(error))
+    )
+    .catch(error => this.catchExpiredToken(error));
+  }
+
+  private postMessageWithAttachments({ attachments, ... messageWithoutAttachments }: Message) {
```

Comment:
here you have `private` and in other places it isn't there. I prefer `private` on all along with the return type (even if void).